spec/libraries/duplicator_spec.rb

```diff
@@ -0,0 +1,478 @@
+      new_children = []
+
+      @children.each do |child|
+        new_child = duplicator.duplicate(child)
+        new_children << new_child unless new_child.nil?
+      end
+
+      # example if object has no accessor for the children
+      # self is already duplicated by Duplicator. Find the duplicated version
+      duplicator.duplicate(self).instance_variable_set(:@children, new_children)
+    end
+
+    def ==(other)
+      self.class == other.class && state == other.state
+    end
+
+    protected
+
+    def state
+      [@id, @children]
+    end
+  end
+
+  def create_complex_objects
+    @s1 = SimpleObject.new(1)
+    @s2 = SimpleObject.new(2)
+    @s3 = SimpleObject.new(3)
+
+    # setup 2 objects with overlapping children
+    @c1 = ComplexObject.new(11, [@s1, @s2])
+    @c2 = ComplexObject.new(12, [@s1])
+
+    # setup an even more complicated object
+    @c3 = ComplexObject.new(13, [@c1, @c2, @s3])
+  end
+
+  def create_cyclic_graph
+    #
+    #       ------> c3 ---
+    #       |       ^    |
+    # c1 -> c2      |    -> c5
+    #       |       |       |
+    #       ------> c4 <-----
+    #               |
+    #               -----> s1
+    #
+    @s1 = SimpleObject.new(1)
+    @c4 = ComplexObject.new(14, [])      # assign children later
+    @c5 = ComplexObject.new(15, [@c4])
+    @c3 = ComplexObject.new(13, [@c5])
+    @c2 = ComplexObject.new(12, [@c3, @c4])
+    @c1 = ComplexObject.new(11, [@c2])
+
+    @c4.instance_variable_set(:@children, [@s1, @c3])   # create cycle
+  end
+
+  def create_second_cyclic_graph
+    #
+    #      ----> c22 <--
+    #      |      |    |
+    # c21 --      |    |
+    #      |      |--> |
+    #      ----------> c23
+    #
+    @c22 = ComplexObject.new(22, [])  # assign children later
+    @c23 = ComplexObject.new(23, [@c22])
+    @c21 = ComplexObject.new(21, [@c22, @c23])
+
+    @c22.instance_variable_set(:@children, [@c23])
+  end
+
+  class SimpleActiveRecord < ActiveRecord::Base
+    def initialize_duplicate(_duplicator)
+    end
+
+    def ==(other)
+      self.class == other.class && state == other.state
+    end
+
+    protected
+
+    def state
+      [@data]
+    end
+  end
+
+  temporary_table(:simple_active_records) do |t|
+    t.integer :data
+  end
+
+  class ComplexActiveRecord < ActiveRecord::Base
+    has_and_belongs_to_many :children, class_name: "ComplexActiveRecord", foreign_key: "children_id", join_table: :children_parents
+    has_and_belongs_to_many :parents, class_name: "ComplexActiveRecord", foreign_key: "parent_id", join_table: :children_parents
+    #has_many :parents, class_name: "ComplexActiveRecord", foreign_key: "parent_id", join_table: :children_parents
+    #has_many :child, class_name: "ComplexActiveRecord", foreign_key: "child_id", join_table: :children_parents
+
+    def initialize_duplicate(duplicator)
+    end
+  end
+
+  temporary_table(:complex_active_records) do |t|
+    t.integer :data
+  end
+
+  temporary_table(:children_parents) do |t|
```

Comment:
Yes, but I've already specified the join table, foreign key ID names and class names in the has_and_belongs_to_many.

Is the support spec ignoring it for some reason?